### ISSUE TITLE
L10 rule in pyproject.toml is ignored

### ISSUE THREAD

**Issue by jan-benisek** [sqlfluff/sqlfluff#2232]
## Expected Behaviour

So I am configuring the linter using `pyproject.toml`. I want all key words UPPERCASE.
```
[sqlfluff.rules.L010]
capitalisation_policy = "upper"
```
When I run lint and fix
```
sqlfluff lint --dialect 'redshift' foo/bar/test.sql
sqlfluff fix --force --dialect 'redshift' foo/bar/test.sql
```

on this file:
```sql
select a+b  as foo,
c as bar from my_table;
```

I expect this:
```sql
SELECT
    c AS bar,
    a + b AS foo
FROM my_table;
```

## Observed Behaviour
The linter gives no warning about rule L10
```
sqlfluff lint --dialect 'redshift' foo/bar/test.sql
sqlfluff fix --force --dialect 'redshift' foo/bar/test.sql
```

And this is the result
```sql
select
    c as bar,
    a + b as foo
from my_table;
```

**IF** I use `.sqlfluff` with this:
```
[sqlfluff:rules:L010]
capitalisation_policy = upper
```
everything works as it should

## Steps to Reproduce
See above

## Dialect
None

## Version
sqlfluff, version 0.9.0

## Configuration
See above

**Comment by jpy-git** [comment]
Pretty sure for `pyproject.toml` you should be using:
```
[tool.sqlfluff.rules.L010]
capitalisation_policy = "upper"
```

the `tool.` prefix is standard for `pyproject.toml` and there's also a note on it in the docs: https://docs.sqlfluff.com/en/stable/configuration.html

**Comment by jan-benisek** [comment]
That worked, thanks!
The documentation is not very clear in this regard. Otherwise great job, starting to really like the library.

**Comment by tunetheweb** [comment]
Any suggestions on how to make the documentation clearer @jan-benisek ?

**Comment by jan-benisek** [comment]
@tunetheweb 
Firstly, I want to say they this is a great tool, the documentation is already better than average and I know that supporting open source is a tough work. I do not criticise, only trying to help you.

I do not know whether this is possible (and it is probably late), but having the same syntax (dots and "tools") for `.sqlfluff` and `pyproject.toml` would be great.

It would be nice to receive an error if certain line in `pyproject.toml` is ignored. In my case, this:
```
[sqlfluff.rules.L010]
capitalisation_policy = "upper"
```
Gave no error, which added to the confusion.

Lastly, from the documentation it is not really clear which rules belong where. When I am looking at the [default configuration](https://docs.sqlfluff.com/en/stable/configuration.html#default-configuration), I am wondering:

- `output_line_length` - nowhere in documentation (I used search) is explained what this means (example, there are more fields like this)
- I could not find what are all the possible fields in the `configuration` file and what are the options. Sure, there is the example with the default values and I could probably search _some_ of the attributes (see previous point), but it makes my life hard.
- What are the levels and where do they come from?
    - For example `[sqlfluff:templater]` - what are other possible second levels (I see `indentation` and `rules`. 
    - The example with `pyproject.toml` has `sqlfluff.core` level - do I need to use it? Why is it not in the default example? Can I use just `[tool.sqlfluff]`?
- The same settings can be configured in multiple places, how do I choose which goes where? Does one takes precedence over the other? Documentation does not mention that, or I could not find it.
```
[tool.sqlfluff.indentation]
indent_unit = 'tab'

[tool.sqlfluff.rules]
indent_unit = 'tab'
```

**Comment by tunetheweb** [comment]
Wow lot there! 😀 I was hoping there was a simple thing we could do to make the toml set up more obvious, but seems like there's some deeper concerns here!

> @tunetheweb Firstly, I want to say they this is a great tool, the documentation is already better than average and I know that supporting open source is a tough work. I do not criticise, only trying to help you.

Yup, get that, whatever we can do to improve is good for everyone

> I do not know whether this is possible (and it is probably late), but having the same syntax (dots and "tools") for `.sqlfluff` and `pyproject.toml` would be great.

I think it is covered here: https://docs.sqlfluff.com/en/stable/configuration.html#default-configuration as @jpy-git stated. And it has some examples - though not the full config. To be honest, I'd be against repeating the whole thing multiple times so think what we currently have should be sufficient. But you clearly missed it, so up for hearing any suggestions to improve this. But do think a full repeat is a bit too much and will ultimately be more confusing, plus more effort to maintain (which goes to some of your points below).

> It would be nice to receive an error if certain line in `pyproject.toml` is ignored. In my case, this:
> 
> ```
> [sqlfluff.rules.L010]
> capitalisation_policy = "upper"
> ```
> 
> Gave no error, which added to the confusion.

OK. Not sure how feasible that is but agree it would be nice. Not sure if this is a change in sqlfluff, or toml (we use the `toml` Python package to read this). Imagine it's the same for typos in `sqlfluffini` files to be honest. There's also a risk here as I often include upcoming config waiting on next SQLFluff release if I (as a maintainer) know there's new config coming and I don't want old versions to fail because of "unrecognised" config.

> Lastly, from the documentation it is not really clear which rules belong where. When I am looking at the [default configuration](https://docs.sqlfluff.com/en/stable/configuration.html#default-configuration), I am wondering:
> 
> * `output_line_length` - nowhere in documentation (I used search) is explained what this means (example, there are more fields like this)

This could be clarified I agree, especially the difference between that and `sqlfluff:rules:max_line_length`. It seems to me (from a very quick search of the code base) that former is used for outputing error message, while the later is used to enforce SQL rules. But not obvious. We should improve this. A comment in the default config would be good for any non-obvious config — though that's in the eye of the beholder — and also want to avoid a really huge default config file tha's more verbose than it needs to be.

> * I could not find what are all the possible fields in the `configuration` file and what are the options. Sure, there is the example with the default values and I could probably search _some_ of the attributes (see previous point), but it makes my life hard.
> * What are the levels and where do they come from?
>   
>   * For example `[sqlfluff:templater]` - what are other possible second levels (I see `indentation` and `rules`.
>   * The example with `pyproject.toml` has `sqlfluff.core` level - do I need to use it? Why is it not in the default example? Can I use just `[tool.sqlfluff]`?

As can be seen from the default config, there's basically the following:
- `[sqlfluff]` (also known as `sqlfluff.core`). Note sure if/why toml needs the explict `.core`. mThat's a good question
- `[sqlfluff:indentation]` - agree this should just be moved to the core section IMHO
- `[sqlfluff:templater]` - this is covered in the [templating config documentation](https://docs.sqlfluff.com/en/stable/configuration.html#jinja-templating-configuration)
- `[sqlfluff:rules]` - this is covered in the [rules documentation](https://docs.sqlfluff.com/en/stable/rules.html).

The final part is also covered in the default config:

```
# Some rules can be configured directly from the config common to other rules.
[sqlfluff:rules]
tab_space_size = 4
max_line_length = 80
indent_unit = space
comma_style = trailing
allow_scalar = True
single_table_references = consistent
unquoted_identifiers_policy = all

# Some rules have their own specific config.
[sqlfluff:rules:L007]  # Keywords
operator_new_lines = after
...
```

Maybe the top common rules config should be moved to the `core` section as agree there's no clear distinction between what's `core` and what's `common rules` config, though also nt against keeping them where they are.

> * The same settings can be configured in multiple places, how do I choose which goes where? Does one takes precedence over the other? Documentation does not mention that, or I could not find it.
> 
> ```
> [tool.sqlfluff.indentation]
> indent_unit = 'tab'
> 
> [tool.sqlfluff.rules]
> indent_unit = 'tab'
> ```

I can't see `indent_unit` anywhere under `indentation`. Agree that's somewhat confusing since we have a specific `sqlfluff.indentation` section and this is clearly an indent config, but the sample config should at least tell you where this `indent_unit` belongs. Can you let me know if you did see it documented under `sqlfluff.indentation` somewhere in case I've missed something? Or if this is just an assumption you made (not a crazy assumption given the name I admit!).

**Comment by jan-benisek** [comment]
> I think it is covered here: https://docs.sqlfluff.com/en/stable/configuration.html#default-configuration as @jpy-git stated. And it has some examples - though not the full config. To be honest, I'd be against repeating the whole thing multiple times so think what we currently have should be sufficient. But you clearly missed it, so up for hearing any suggestions to improve this. But do think a full repeat is a bit too much and will ultimately be more confusing, plus more effort to maintain (which goes to some of your points below).

Agree, no need to repeat. Once place with example is more than enough. The point I have, and will iterate on it later, is that I am missing full description of what are all the possible values in the config and to which group they belong.

> This could be clarified I agree, especially the difference between that and sqlfluff:rules:max_line_length. It seems to me (from a very quick search of the code base) that former is used for outputing error message, while the later is used to enforce SQL rules. But not obvious. We should improve this. A comment in the default config would be good for any non-obvious config — though that's in the eye of the beholder — and also want to avoid a really huge default config file tha's more verbose than it needs to be.

Exactly, not obvious and you cannot search for it anywhere on the docs. Search through code base is probably not an option if I need a quick fix.
I think that the config like this is great and easy to use. I could imagine having a link (or a sub-page) where everything is explained in detail.

> As can be seen from the default config, there's basically the following:

That's the thing, I see the default config, but is there anything else which I could tweak? A simple sentence to clarify this can be helpful "These are all the fields which can be configured at this point". And as mentioned before, adding "Here (link) is a full description of those fields and categories" would be great.

> Maybe the top common rules config should be moved to the core section as agree there's no clear distinction between what's core and what's common rules config, though also nt against keeping them where they are.

This would really help, having one clear option where to put general settings.

> I can't see indent_unit anywhere under indentation. Agree that's somewhat confusing since we have a specific sqlfluff.indentation section and this is clearly an indent config, but the sample config should at least tell you where this indent_unit belongs. Can you let me know if you did see it documented under sqlfluff.indentation somewhere in case I've missed something? Or if this is just an assumption you made (not a crazy assumption given the name I admit!).

Looking at it, I probably put it there by myself. I was confused what belongs where and that was the result. Which brings me back to the previous point - having clear description what goes where.

**Comment by jan-benisek** [comment]
> I think it is covered here: https://docs.sqlfluff.com/en/stable/configuration.html#default-configuration as @jpy-git stated. And it has some examples - though not the full config. To be honest, I'd be against repeating the whole thing multiple times so think what we currently have should be sufficient. But you clearly missed it, so up for hearing any suggestions to improve this. But do think a full repeat is a bit too much and will ultimately be more confusing, plus more effort to maintain (which goes to some of your points below).

Agree, no need to repeat. Once place with example is more than enough. The point I have, and will iterate on it later, is that I am missing full description of what are all the possible values in the config and to which group they belong.

> This could be clarified I agree, especially the difference between that and sqlfluff:rules:max_line_length. It seems to me (from a very quick search of the code base) that former is used for outputing error message, while the later is used to enforce SQL rules. But not obvious. We should improve this. A comment in the default config would be good for any non-obvious config — though that's in the eye of the beholder — and also want to avoid a really huge default config file tha's more verbose than it needs to be.

Exactly, not obvious and you cannot search for it anywhere on the docs. Search through code base is probably not an option if I need a quick fix.
I think that the config like this is great and easy to use. I could imagine having a link (or a sub-page) where everything is explained in detail.

> As can be seen from the default config, there's basically the following:

That's the thing, I see the default config, but is there anything else which I could tweak? A simple sentence to clarify this can be helpful "These are all the fields which can be configured at this point". And as mentioned before, adding "Here (link) is a full description of those fields and categories" would be great.

> Maybe the top common rules config should be moved to the core section as agree there's no clear distinction between what's core and what's common rules config, though also nt against keeping them where they are.

This would really help, having one clear option where to put general settings.

> I can't see indent_unit anywhere under indentation. Agree that's somewhat confusing since we have a specific sqlfluff.indentation section and this is clearly an indent config, but the sample config should at least tell you where this indent_unit belongs. Can you let me know if you did see it documented under sqlfluff.indentation somewhere in case I've missed something? Or if this is just an assumption you made (not a crazy assumption given the name I admit!).

Looking at it, I probably put it there by myself. I was confused what belongs where and that was the result. Which brings me back to the previous point - having clear description what goes where.

**Comment by keatmin** [comment]
Got the same issue trying to configure `pyproject.toml` and had to look up the issue. It's documented but sometimes I process things better if there's an example.  My suggestion is having a simple example (side by side) on how things should be for configs in `pyproject.toml` vs others can definitely help. 
```
pyproject vs others
[tool.sqlfluff.rules.L007] vs [sqlfluff:rules:L007]
```
